File: 22. Generate Parentheses/22. Generate Parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
  const ans = [];
  const curr = [];

  const backtrack = (start, end) => {
    if (curr.length === 2 * n) {
      ans.push(curr.join(""));
      return;
    }

    if (start < n) {
      curr.push("(");
      backtrack(start + 1, end);
      curr.pop();
    }

    if (end < start) {
      curr.push(")");
      backtrack(start, end + 1);
      curr.pop();
    }
  };

  backtrack(0, 0);
  return ans;
};

File: 22. Generate Parentheses/22. Generate Parentheses.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans, curr = [], []

        def backtrack(start, end):
            if len(curr) == n * 2:
                ans.append("".join(curr[:]))
                return

            if start < n:
                curr.append("(")
                backtrack(start + 1, end)
                curr.pop()

            if end < start:
                curr.append(")")
                backtrack(start, end + 1)
                curr.pop()

        backtrack(0, 0)
        return ans
